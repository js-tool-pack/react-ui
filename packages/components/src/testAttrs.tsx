import { fireEvent, render } from '@testing-library/react';
import type { PropsBase } from '@pkg/shared';
import React from 'react';

export function testAttrs<
  T extends { attrs?: Partial<React.HTMLAttributes<HTMLElement>> },
>(component: React.FC<T>, displayName?: string) {
  const name = displayName || component.displayName || component.name;
  test(name ? `${name} attrs` : `attrs`, () => {
    const onClick = jest.fn();

    const FC = component as React.FC<PropsBase>;
    const { container } = render(
      <FC
        attrs={{
          style: { width: '20px' },
          className: 'foo',
          role: 'note',
          onClick,
        }}
      />,
    );

    expect(container.firstChild).toHaveClass('foo');
    expect(container.firstChild).toHaveAttribute('role', 'note');
    expect(container.firstChild).toHaveStyle({ width: '20px' });

    expect(onClick).not.toBeCalled();
    fireEvent.click(container.firstChild!);
    expect(onClick).toBeCalled();
  });
}
