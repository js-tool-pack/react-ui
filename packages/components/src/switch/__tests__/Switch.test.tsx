import { fireEvent, render } from '@testing-library/react';
import { testAttrs } from '~/testAttrs';
import { Switch } from '..';

describe('Switch', () => {
  testAttrs(Switch);

  test('snap', () => {
    const { container } = render(<Switch />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('checked', () => {
    const { container } = render(<Switch checked />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('content', () => {
    const { container } = render(
      <Switch uncheckedChildren="false" checkedChildren="true" />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('disabled', () => {
    const { container } = render(<Switch disabled />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('icon', () => {
    expect(
      render(<Switch uncheckedIcon={2} checkedIcon={1} />).container.firstChild,
    ).toMatchSnapshot();
    expect(
      render(<Switch uncheckedIcon={2} checkedIcon={1} checked />).container
        .firstChild,
    ).toMatchSnapshot();
  });

  test('loading', () => {
    expect(render(<Switch loading />).container.firstChild).toMatchSnapshot();
  });

  test('onChange', () => {
    const onChange = jest.fn();
    const { container } = render(<Switch onChange={onChange} />);

    expect(onChange).not.toBeCalled();
    fireEvent.click(container.firstChild!);
    expect(onChange).toBeCalled();
  });
});
