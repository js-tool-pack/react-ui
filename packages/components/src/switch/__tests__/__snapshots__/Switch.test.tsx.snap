// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch checked 1`] = `
<button
  aria-checked="true"
  class="t-switch t--size-m t-switch--checked"
  role="switch"
>
  <div
    class="t-switch__handle"
  />
  <div
    class="t-switch__slider"
  >
    <div
      class="t-switch__checked"
    />
    <div
      class="t-switch__unchecked"
    />
  </div>
</button>
`;

exports[`Switch content 1`] = `
<button
  aria-checked="false"
  class="t-switch t--size-m t-switch--unchecked"
  role="switch"
>
  <div
    class="t-switch__handle"
  />
  <div
    class="t-switch__slider"
  >
    <div
      class="t-switch__checked"
    >
      true
    </div>
    <div
      class="t-switch__unchecked"
    >
      false
    </div>
  </div>
</button>
`;

exports[`Switch disabled 1`] = `
<button
  aria-checked="false"
  aria-disabled="true"
  class="t-switch t--size-m t-switch--unchecked t-switch--disabled"
  disabled=""
  role="switch"
>
  <div
    class="t-switch__handle"
  />
  <div
    class="t-switch__slider"
  >
    <div
      class="t-switch__checked"
    />
    <div
      class="t-switch__unchecked"
    />
  </div>
</button>
`;

exports[`Switch icon 1`] = `
<button
  aria-checked="false"
  class="t-switch t--size-m t-switch--unchecked"
  role="switch"
>
  <div
    class="t-switch__handle"
  >
    <i
      class="t-icon t-switch__icon"
      role="img"
    >
      2
    </i>
  </div>
  <div
    class="t-switch__slider"
  >
    <div
      class="t-switch__checked"
    />
    <div
      class="t-switch__unchecked"
    />
  </div>
</button>
`;

exports[`Switch icon 2`] = `
<button
  aria-checked="true"
  class="t-switch t--size-m t-switch--checked"
  role="switch"
>
  <div
    class="t-switch__handle"
  >
    <i
      class="t-icon t-switch__icon"
      role="img"
    >
      1
    </i>
  </div>
  <div
    class="t-switch__slider"
  >
    <div
      class="t-switch__checked"
    />
    <div
      class="t-switch__unchecked"
    />
  </div>
</button>
`;

exports[`Switch loading 1`] = `
<button
  aria-checked="false"
  class="t-switch t--size-m t-switch--unchecked t-switch--loading"
  role="switch"
>
  <div
    class="t-switch__handle"
  >
    <i
      class="t-icon t-switch__icon"
      role="img"
    >
      <svg
        class="t-loading-icon"
        viewBox="0 0 66 66"
      >
        <circle
          class="t-loading-icon__path"
          cx="33"
          cy="33"
          fill="none"
          r="30"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-width="6"
        />
      </svg>
    </i>
  </div>
  <div
    class="t-switch__slider"
  >
    <div
      class="t-switch__checked"
    />
    <div
      class="t-switch__unchecked"
    />
  </div>
</button>
`;

exports[`Switch snap 1`] = `
<button
  aria-checked="false"
  class="t-switch t--size-m t-switch--unchecked"
  role="switch"
>
  <div
    class="t-switch__handle"
  />
  <div
    class="t-switch__slider"
  >
    <div
      class="t-switch__checked"
    />
    <div
      class="t-switch__unchecked"
    />
  </div>
</button>
`;
